pkg/agent/proxy/proxier.go

```diff
@@ -212,47 +212,31 @@ func getBindingProtoForIPProto(endpointIP string, protocol corev1.Protocol) bind
 	return bindingProtocol
 }
 
-// removeEndpoint removes flows for the given Endpoint from the data path if these flows are no longer
+// removeStaleEndpoints removes flows for the given Endpoints from the data path if these flows are no longer
 // needed by any Service. Endpoints from different Services can have the same characteristics and thus
-// can share the same flows. removeEndpoint must be called whenever an Endpoint is no longer used by a
-// given Service. If the Endpoint is still referenced by any other Services, no flow will be removed.
+// can share the same flows. removeStaleEndpoints must be called whenever Endpoints are no longer used by a
+// given Service. If the Endpoints are still referenced by any other Services, no flow will be removed.
 // The method only returns an error if a data path operation fails. If the flows are successfully
-// removed from the data path, the method returns true. Otherwise, if the flows are still needed for
-// other Services, it returns false.
-func (p *proxier) removeEndpoint(endpoint k8sproxy.Endpoint, protocol binding.Protocol) (bool, error) {
-	key := endpointKey(endpoint, protocol)
-	count := p.endpointReferenceCounter[key]
-	if count == 1 {
-		if err := p.ofClient.UninstallEndpointFlows(protocol, endpoint); err != nil {
-			return false, err
+// removed from the data path, the method returns nil.
+func (p *proxier) removeStaleEndpoints(staleEndpoints map[string]k8sproxy.Endpoint, ipProtocol corev1.Protocol) error {
+	var endpointsToRemove []k8sproxy.Endpoint
+	bindingProtocol := getBindingProtoForIPProto(p.isIPv6, ipProtocol)
+
+	for _, endpoint := range staleEndpoints {
+		key := endpointKey(endpoint, bindingProtocol)
+		count := p.endpointReferenceCounter[key]
+		if count == 1 {
+			endpointsToRemove = append(endpointsToRemove, endpoint)
+			klog.V(2).InfoS("Endpoint will be removed", "Endpoint", endpoint.String(), "Protocol", bindingProtocol)
+		} else {
+			p.endpointReferenceCounter[key] = count - 1
```

Comment:
According to the code, if `UninstallEndpointFlows` fails, the endpoint's counter is still deducted, then next attempt would deduct twice and mess up the counter, leading to the endpoint being removed unexpectedly.

Reply:
Update reference counter after `UninstallEndpointFlows` succeeds.